### PR TITLE
C#: Refactor pattern extraction

### PR DESCRIPTION
**Pull request by tamasvajk**
* separated the classes into dedicated files
* moved all pattern classes to a dedicated folder
* deleted duplicated pattern extraction logic (`IsPattern` vs `CasePattern`).

### CODE REVIEW

**Review by hvitved — APPROVED** [review]
Much better, thanks!

**Comment by tamasvajk** [comment]
I just started a [differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/628/). Let's wait for it to finish before I merge this.

**Pull request merged**
